py-polars/polars/functions.py

```diff
@@ -1,85 +1,125 @@
+    ...
 
-    Parameters
-    ----------
-    url
-        str or unicode.
 
-    Returns
-    -------
-    isurl : bool
-        If `url` has a valid protocol return True otherwise False.
-    """
-    try:
-        return parse_url(url).scheme in _VALID_URLS
-    except Exception:
-        return False
+@overload
+def _prepare_file_arg(
+    file: Union[str, List[str], TextIO, Path, BinaryIO], **kwargs: Any
+) -> ContextManager[Union[str, BinaryIO, List[str], List[BinaryIO]]]:
+    ...
 
 
 def _prepare_file_arg(
-    file: Union[str, TextIO, Path, BinaryIO]
-) -> Union[str, TextIO, Path, BinaryIO]:
+    file: Union[str, List[str], TextIO, Path, BinaryIO], **kwargs: Any
+) -> ContextManager[Union[str, BinaryIO, List[str], List[BinaryIO]]]:
     """
     Utility for read_[csv, parquet]. (not to be used by scan_[csv, parquet]).
+    Returned value is always usable as a context.
+
+    A `StringIO`, `BytesIO` file is returned as a `BytesIO`
+    A local path is returned as a string
+    An http url is read into a buffer and returned as a `BytesIO`
 
-    Does one of:
-        - A path.Path object is converted to a string.
-        - A raw file on the web is downloaded into a buffer.
+    When fsspec is installed, except for `StringIO`, `BytesIO` and local
+    uncompressed files, the file is opened with `fsspec.open(file, **kwargs)`,
+    in which case, the compression is inferred.
     """
-    if _is_url(file):
-        if isinstance(file, Path):
-            file = str(file)
 
-        if isinstance(file, str) and file.startswith("http"):
-            file = _process_http_file(file)
-    else:
-        if isinstance(file, StringIO):
-            return io.BytesIO(file.read().encode("utf8"))
-        elif isinstance(file, BytesIO):
-            return file
-    return file
+    # Small helper to use a string as context
+    @contextmanager
+    def managed_file(file: Union[str, List[str]]) -> Iterator[Union[str, List[str]]]:
+        try:
+            yield file
+        finally:
+            pass
+
+    if isinstance(file, StringIO):
+        return io.BytesIO(file.read().encode("utf8"))
+    if isinstance(file, BytesIO):
+        return file
+    if isinstance(file, Path):
+        return managed_file(str(file))
+    if isinstance(file, str):
+        if WITH_FSSPEC:
+            compressed = infer_compression(file) is not None
+            local = infer_storage_options(file)["protocol"] == "file"
+            if local and not compressed:
+                return managed_file(make_path_posix(file))
+            compression = kwargs.pop("compression", "infer")
+            return fsspec.open(file, compression=compression, **kwargs)
+        if file.startswith("http"):
+            return _process_http_file(file)
+        return managed_file(file)
+    if isinstance(file, list) and bool(file) and all(isinstance(f, str) for f in file):
+        if WITH_FSSPEC:
+            compressed = any(infer_compression(f) is not None for f in file)
```

Comment:
@ritchie46 Compression and protocol is checked for all files in the list. we could maybe check for the first one only? I can't really think of any case where these would be mixed.